### PR TITLE
adding ability to set checked state for menu item at registration

### LINKED ISSUE #14986: Show checkmarks in top level menu to indicate activity of a toggle
- VSCode Version: 1.8.0 - insider (dbf3ba7)
- OS Version: Windows 10

-- Sorry for the English, I used Google Translator --

Place a check in some menu to see if it is active or not. Similar to the Auto Save File> Auto Save

Mainly for options that do not change the text in the menu.

Example: Toggle Word Wrap, Toggle Render WhiteSpace, Toggle Full Screen, Toggle Developer Tools, among others.

### LINKED ISSUE #56885: Make visible in view menu the toggle state
Issue Type: <b>Feature Request</b>

In the View menu the status of the various "Toggle..." entries is not visible. Please add a tickmark or something to make their state visible. Thanks!

VS Code version: Code 1.26.1 (493869ee8e8a846b0855873886fc79d480d342de, 2018-08-16T18:38:57.434Z)
OS version: Windows_NT x64 10.0.17134

### LINKED ISSUE #58422: Provide a mechanism to set a menu entry to "checked"
It looks like this is a hack only for one menu item: https://github.com/Microsoft/vscode/blob/677b68da48fb4e1f518801fd06fba2200fdc6e41/src/vs/workbench/browser/parts/titlebar/menubarControl.ts#L491

There should be a general solution to set a menu entry to be checked or not. This enables: https://github.com/Microsoft/vscode/issues/46807 and https://github.com/Microsoft/vscode/issues/14986

### PR DESCRIPTION
**Pull request by sbatten**
resolves #58422
resolves #56885
resolves #14986

word wrap has a transient state which is what is controlled by the action (not the setting), so I haven't done that one yet.

Can also look into #46807 as well.

### CODE REVIEW

**Review by bpasero — APPROVED** [review]
This works great. It is a bit weird though to have our various "Toggle..." actions in the menu have a check box. I would rather expect those commands to be named "Show..." and then the check box indicates if the thing is shown or not. Or probably the commands should change to "Hide..." to indicate the opposite. We should probably create a follow up issue to fix that.

File: src/vs/platform/actions/common/actions.ts [hunk 1]
```diff
@@ -27,6 +27,7 @@ export interface IBaseCommandAction {
 export interface ICommandAction extends IBaseCommandAction {
 	iconLocation?: { dark: URI; light?: URI; };
 	precondition?: ContextKeyExpr;
+	effectToggled?: ContextKeyExpr;
```
Comment (jrieken) [inline src/vs/platform/actions/common/actions.ts]: This name is weird. They not `toggled` or `checked`?

File: src/vs/platform/actions/common/actions.ts [hunk 2]
```diff
@@ -246,6 +246,10 @@ export class MenuItemAction extends ExecuteCommandAction {
 		typeof item.title === 'string' ? super(item.id, item.title, commandService) : super(item.id, item.title.value, commandService);
 		this._cssClass = undefined;
 		this._enabled = !item.precondition || contextKeyService.contextMatchesRules(item.precondition);
+		if (item.effectToggled && contextKeyService.contextMatchesRules(item.effectToggled)) {
+			this._checked = true;
+		}
```
Comment (jrieken) [inline src/vs/platform/actions/common/actions.ts]: better, clearer? `this._check = item.effectToggled && contextKeyService.contextMatchesRules(item.effectToggled)`

**Review by jrieken — APPROVED** [review]

**Comment by eamodio** [comment]
@sbatten @jrieken can extensions use this?

**Comment by jrieken** [comment]
not yet

**Pull request merged**